students/MzKhan/lesson04/mailroom2.py

```diff
@@ -0,0 +1,217 @@
+    # parm: list
+    """
+    title = "{:24} | {:12} | {:10} | {:16}"
+    print(title.format('Donor Name','Total Given','Num Gifts','Average Gift'))
+    strf_format = "{:12} {:12} $  {:10.2f} {:10d}   ${:10.2f}"
+    print(67*('-'))
+    report_list = calculate_total_gift(donors_data)
+    for donor in sorted(report_list, key = igetter(2), reverse = True):
+        print(strf_format.format(*donor))
+
+def calculate_total_gift(donors_data):
+    """
+    This method creates a new list from the global donors' list
+    where each item of the list is a list of
+    the donor's first name, last name, total donation, number of donations,
+    and an average donation
+    parm: list
+    return: list
+    """
+    new_list = []
+    for donor in donors_data:
+        first,last = donor[0],donor[1]
+        total_donation, num_of_gifts = sum(donor[2:]), len(donor)-2
+        avg_gift = total_donation/num_of_gifts
+        new_list.append([first,last,total_donation,num_of_gifts,avg_gift])
+    return new_list
+
+def quit():
+    """
+    This method exits out from the user interaction.
+    """
+    print("Thank you. Have a nice day:")
+    sys.exit()
+
+def send_letter_everyone(donors_data):
+    '''
+    This method generates the thank you letters for each donor in the list
+    parm: list
+    '''
+    date_format = '{:%m-%d-%Y}'.format(datetime.datetime.now())
+    for donor in donors_data:
+        _donor = []
+        _donor.extend(donor[0:2])
+        _donor.extend([donor[-1],sum(donor[2:]),len(donor)-2])
+        letter=letter_format().format(*_donor)
+        file_name = '_'.join(_donor[0:2])+"_"+date_format+'.txt'
+        write_a_letter(file_name,letter)
+
+def write_a_letter(filename, content):
+    '''This method creates a new file in the Letters folder and writes to it.'''
+    folder = "Letters"
+    if not os.path.exists(folder):
+        os.mkdir(folder)
+    with open(os.path.join(folder,filename),'w+') as out:
+        out.write(content)
+        out.close()
+
+def letter_format():
+    """
+    return: string
+    """
+    message = """Dear {} {},
+
+    Thank you so much for your kind donation of ${:.2f}. With that you have
+    generously donated a total amount of ${:.2f} in your {} donation(s).
+    We must ensure you that your donations will be put to a very good use.
+
+                                                        Sincerely,
+
+                                                        -Team """
+    return message
+
+
+if __name__ == "__main__":
+
+    #Main program flow control and the interaction with the user.
+    options = {1:thank_you,2:print_report,3:send_letter_everyone,4:quit}
+    while True:
+        user_response = prompt_user()
+        if user_response == 4:
+            options[user_response]()
+        elif user_response in options:
+            options[user_response](donors_list)
+        else:
+            print("Invalid Response Please Try Again :)")
```

Comment:
you don't need to pass in `donors_list` since its a global variable (module level)

you can simplify this to:
```
while True:
    if user_response in options:
        options[user_response]()
    else:
        print('invalid')
```